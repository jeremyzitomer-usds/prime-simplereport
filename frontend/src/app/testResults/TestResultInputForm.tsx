import React from "react";

import RadioGroup from "../commonComponents/RadioGroup";
import Button from "../commonComponents/Button";
import { COVID_RESULTS, TEST_RESULT_DESCRIPTIONS } from "../constants";
import { TestResult } from "../testQueue/QueueItem";

interface Props {
  queueItemId: string;
  testResultValue: TestResult | undefined;
  onChange: (value: TestResult | undefined) => void;
  onSubmit: () => void;
}

const TestResultInputForm: React.FC<Props> = ({
  queueItemId,
  testResultValue,
  onSubmit,
  onChange,
}) => {
  const onResultChange = (event: React.FormEvent<HTMLInputElement>) => {
    const value = (event.target as HTMLInputElement).value as TestResult;
    onChange(value);
  };

  const onResultClick = (event: React.FormEvent<HTMLInputElement>) => {
    const value = (event.target as HTMLInputElement).value as TestResult;
    if (value === testResultValue) {
      onChange(undefined);
    }
  };

  const onResultSubmit = (event: React.FormEvent<HTMLButtonElement>) => {
    event.preventDefault();
    onSubmit();
  };

  return (
    <form className="usa-form">
      <h4 className="prime-radio__title"> SARS-CoV-2 Results </h4>
      <React.Fragment>
        <RadioGroup
          legend="Test result"
          legendSrOnly
          onClick={onResultClick}
          onChange={onResultChange}
          buttons={[
            {
              value: COVID_RESULTS.POSITIVE,
              label: `${TEST_RESULT_DESCRIPTIONS.POSITIVE} (+)`,
            },
            {
              value: COVID_RESULTS.NEGATIVE,
              label: `${TEST_RESULT_DESCRIPTIONS.NEGATIVE} (-)`,
            },
            {
              value: COVID_RESULTS.INCONCLUSIVE,
              label: `${TEST_RESULT_DESCRIPTIONS.UNDETERMINED}`,
            },
          ]}
          name={`covid-test-result-${queueItemId}`}
          selectedRadio={testResultValue}
        />
        <div className="prime-test-result-submit">
          <Button
            onClick={onResultSubmit}
            type="submit"
            disabled={!testResultValue}
            variant="outline"
            label="Submit"
          />
        </div>
      </React.Fragment>
    </form>
  );
};

export default TestResultInputForm;
