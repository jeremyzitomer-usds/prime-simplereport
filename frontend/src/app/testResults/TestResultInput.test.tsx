import React from "react";
import renderer from "react-test-renderer";
import { render, fireEvent, screen } from "@testing-library/react";
import TestResultInputForm from "./TestResultInputForm";

jest.mock("uuid");

describe("TestResultInputForm", () => {
  it("should render with a value", () => {
    const component = renderer.create(
      <TestResultInputForm
        queueItemId={"5d315d18-82f8-4025-a051-1a509e15c880"}
        testResultValue="POSITIVE"
        onChange={jest.fn()}
        onSubmit={jest.fn()}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render without a value", () => {
    const component = renderer.create(
      <TestResultInputForm
        queueItemId={"5d315d18-82f8-4025-a051-1a509e15c880"}
        testResultValue={undefined}
        onChange={jest.fn()}
        onSubmit={jest.fn()}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should pass back the result value when clicked", () => {
    const onChange = jest.fn();
    render(
      <TestResultInputForm
        queueItemId={"5d315d18-82f8-4025-a051-1a509e15c880"}
        testResultValue={undefined}
        onChange={onChange}
        onSubmit={jest.fn()}
      />
    );

    fireEvent.click(screen.getByLabelText("Negative (-)"));

    expect(onChange).toBeCalledWith("NEGATIVE");
  });

  it("should remove value when it is already selected", () => {
    const onChange = jest.fn();
    render(
      <TestResultInputForm
        queueItemId={"5d315d18-82f8-4025-a051-1a509e15c880"}
        testResultValue="POSITIVE"
        onChange={onChange}
        onSubmit={jest.fn()}
      />
    );

    fireEvent.click(screen.getByLabelText("Positive (+)"));

    expect(onChange).toBeCalledWith(undefined);
  });

  it("should not submit when there is no value", () => {
    const onSubmit = jest.fn();
    render(
      <TestResultInputForm
        queueItemId={"5d315d18-82f8-4025-a051-1a509e15c880"}
        testResultValue={undefined}
        onChange={jest.fn()}
        onSubmit={onSubmit}
      />
    );

    fireEvent.click(screen.getByText("Submit"));

    expect(onSubmit).toHaveBeenCalledTimes(0);
  });

  it("should submit when there is a value", () => {
    const onSubmit = jest.fn();
    render(
      <TestResultInputForm
        queueItemId={"5d315d18-82f8-4025-a051-1a509e15c880"}
        testResultValue="POSITIVE"
        onChange={jest.fn()}
        onSubmit={onSubmit}
      />
    );

    fireEvent.click(screen.getByText("Submit"));

    expect(onSubmit).toHaveBeenCalledTimes(1);
  });
});
