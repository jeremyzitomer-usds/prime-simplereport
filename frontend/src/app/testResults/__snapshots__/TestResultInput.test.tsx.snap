// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestResultInputForm should render with a value 1`] = `
<form
  className="usa-form"
>
  <h4
    className="prime-radio__title"
  >
     SARS-CoV-2 Results 
  </h4>
  <fieldset
    className="usa-fieldset prime-radios"
  >
    <legend
      className="usa-legend usa-sr-only"
    >
      Test result
    </legend>
    <div
      className="usa-form-group"
    >
      <div
        className="usa-radio"
      >
        <input
          checked={true}
          className="usa-radio__input"
          disabled={false}
          id="radio-test-1_POSITIVE_0"
          name="covid-test-result-5d315d18-82f8-4025-a051-1a509e15c880"
          onChange={[Function]}
          onClick={[Function]}
          type="radio"
          value="POSITIVE"
        />
        <label
          className="usa-radio__label"
          htmlFor="radio-test-1_POSITIVE_0"
        >
          Positive (+)
        </label>
      </div>
      <div
        className="usa-radio"
      >
        <input
          checked={false}
          className="usa-radio__input"
          disabled={false}
          id="radio-test-1_NEGATIVE_1"
          name="covid-test-result-5d315d18-82f8-4025-a051-1a509e15c880"
          onChange={[Function]}
          onClick={[Function]}
          type="radio"
          value="NEGATIVE"
        />
        <label
          className="usa-radio__label"
          htmlFor="radio-test-1_NEGATIVE_1"
        >
          Negative (-)
        </label>
      </div>
      <div
        className="usa-radio"
      >
        <input
          checked={false}
          className="usa-radio__input"
          disabled={false}
          id="radio-test-1_UNDETERMINED_2"
          name="covid-test-result-5d315d18-82f8-4025-a051-1a509e15c880"
          onChange={[Function]}
          onClick={[Function]}
          type="radio"
          value="UNDETERMINED"
        />
        <label
          className="usa-radio__label"
          htmlFor="radio-test-1_UNDETERMINED_2"
        >
          Inconclusive
        </label>
      </div>
    </div>
  </fieldset>
  <div
    className="prime-test-result-submit"
  >
    <button
      className="usa-button usa-button--outline"
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`TestResultInputForm should render without a value 1`] = `
<form
  className="usa-form"
>
  <h4
    className="prime-radio__title"
  >
     SARS-CoV-2 Results 
  </h4>
  <fieldset
    className="usa-fieldset prime-radios"
  >
    <legend
      className="usa-legend usa-sr-only"
    >
      Test result
    </legend>
    <div
      className="usa-form-group"
    >
      <div
        className="usa-radio"
      >
        <input
          checked={false}
          className="usa-radio__input"
          disabled={false}
          id="radio-test-1_POSITIVE_0"
          name="covid-test-result-5d315d18-82f8-4025-a051-1a509e15c880"
          onChange={[Function]}
          onClick={[Function]}
          type="radio"
          value="POSITIVE"
        />
        <label
          className="usa-radio__label"
          htmlFor="radio-test-1_POSITIVE_0"
        >
          Positive (+)
        </label>
      </div>
      <div
        className="usa-radio"
      >
        <input
          checked={false}
          className="usa-radio__input"
          disabled={false}
          id="radio-test-1_NEGATIVE_1"
          name="covid-test-result-5d315d18-82f8-4025-a051-1a509e15c880"
          onChange={[Function]}
          onClick={[Function]}
          type="radio"
          value="NEGATIVE"
        />
        <label
          className="usa-radio__label"
          htmlFor="radio-test-1_NEGATIVE_1"
        >
          Negative (-)
        </label>
      </div>
      <div
        className="usa-radio"
      >
        <input
          checked={false}
          className="usa-radio__input"
          disabled={false}
          id="radio-test-1_UNDETERMINED_2"
          name="covid-test-result-5d315d18-82f8-4025-a051-1a509e15c880"
          onChange={[Function]}
          onClick={[Function]}
          type="radio"
          value="UNDETERMINED"
        />
        <label
          className="usa-radio__label"
          htmlFor="radio-test-1_UNDETERMINED_2"
        >
          Inconclusive
        </label>
      </div>
    </div>
  </fieldset>
  <div
    className="prime-test-result-submit"
  >
    <button
      className="usa-button usa-button--outline usa-button-disabled"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
